apiserver/facades/client/application/application_unit_test.go

```diff
@@ -1926,27 +2225,47 @@ func (s *ApplicationSuite) TestApplicationUpdateSeriesPermissionDenied(c *gc.C)
 }
 
 func (s *ApplicationSuite) TestRemoteRelationBadCIDR(c *gc.C) {
+	defer s.setup(c, state.ModelTypeIAAS).Finish()
+
 	endpoints := []string{"wordpress", "hosted-mysql:nope"}
 	_, err := s.api.AddRelation(params.AddRelation{Endpoints: endpoints, ViaCIDRs: []string{"bad.cidr"}})
 	c.Assert(err, gc.ErrorMatches, `invalid CIDR address: bad.cidr`)
 }
 
 func (s *ApplicationSuite) TestRemoteRelationDisAllowedCIDR(c *gc.C) {
+	defer s.setup(c, state.ModelTypeIAAS).Finish()
+
 	endpoints := []string{"wordpress", "hosted-mysql:nope"}
 	_, err := s.api.AddRelation(params.AddRelation{Endpoints: endpoints, ViaCIDRs: []string{"0.0.0.0/0"}})
 	c.Assert(err, gc.ErrorMatches, `CIDR "0.0.0.0/0" not allowed`)
 }
 
 func (s *ApplicationSuite) TestSetApplicationConfigExplicitMaster(c *gc.C) {
-	s.testSetApplicationConfig(c, model.GenerationMaster)
-}
+	ctrl := s.setup(c, state.ModelTypeCAAS)
+	defer ctrl.Finish()
 
-func (s *ApplicationSuite) TestSetApplicationConfigEmptyUsesMaster(c *gc.C) {
-	s.testSetApplicationConfig(c, "")
-}
+	appCfgSchema, err := caas.ConfigSchema(k8s.ConfigSchema())
+	c.Assert(err, jc.ErrorIsNil)
+	defaults := caas.ConfigDefaults(k8s.ConfigDefaults())
+	appCfgSchema, defaults, err = application.AddTrustSchemaAndDefaults(appCfgSchema, defaults)
+	c.Assert(err, jc.ErrorIsNil)
+
+	appCfg, err := coreconfig.NewConfig(map[string]interface{}{
+		"juju-external-hostname": "value",
+	}, appCfgSchema, nil)
+
+	c.Assert(err, jc.ErrorIsNil)
+	ch := s.expectCharm(ctrl, nil, nil, &charm.Config{
+		Options: map[string]charm.Option{
+			"stringOption": {Type: "string"},
+			"intOption":    {Type: "int", Default: int(123)},
+		},
+	})
+	app := s.expectApplicationWithCharm(ctrl, ch, "postgresql", "kubernetes")
+	app.EXPECT().UpdateCharmConfig(model.GenerationMaster, charm.Settings{"stringOption": "stringVal"})
+	app.EXPECT().UpdateApplicationConfig(appCfg.Attributes(), []string(nil), appCfgSchema, defaults)
+	s.backend.EXPECT().Application("postgresql").Return(app, nil)
 
-func (s *ApplicationSuite) testSetApplicationConfig(c *gc.C, branchName string) {
-	application.SetModelType(s.api, state.ModelTypeCAAS)
```

Comment:
Personally, I'm generally not too keen on extracting the entire body of a test like this. With some exceptions, I don't think it is particularly readable, and unless there's some meaningful reason to extract out a function, can make tests feel rigid

Ideally to avoid duplication tests should be made up of a small number of statements to setup, perform the test, the check with duplication being hidden away behind some declarative statements. 

The degree to which I have done this is certainly questionable. I will re-evaluate these specifically to try and remove as much non-meaningful duplication as possible

Reply:
I believe this can be resolved